internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,264 @@
+		return "", errors.Wrap(err, op)
+	}
+
+	_, err = repo.CreateJob(ctx, j.Name(), code, j.Description())
+	if err != nil && !errors.IsUniqueError(err) {
+		return "", errors.Wrap(err, op)
+	}
+	s.registeredJobs.Store(JobId(jobId), j)
+
+	return JobId(jobId), nil
+}
+
+// UpdateJobNextRun sets the next scheduled run time for the provided jobId
+// to the current database time incremented by the nextRunIn parameter.  If
+// nextRunIn == 0 the job will be available to run immediately.
+//
+// • jobId must be provided and is the id of the job returned by RegisterJob
+//
+// All options are ignored.
+func (s *Scheduler) UpdateJobNextRun(ctx context.Context, jobId JobId, nextRunIn time.Duration, _ ...Option) error {
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if jobId == "" {
+		return errors.New(errors.InvalidParameter, op, "missing job id")
+	}
+	_, ok := s.registeredJobs.Load(jobId)
+	if !ok {
+		return errors.New(errors.InvalidParameter, op, fmt.Sprintf("job %q not registered", jobId))
+	}
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	_, err = repo.UpdateJobNextRun(ctx, string(jobId), nextRunIn)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// Start begins the scheduling loop that will query the repository for jobs to run and
+// run them in a goroutine.
+func (s *Scheduler) Start(ctx context.Context) {
+	if !s.started.CAS(s.started.Load(), true) {
+		s.logger.Debug("scheduler already started, skipping")
+		return
+	}
+
+	s.baseContext, s.baseCancel = context.WithCancel(ctx)
+	go s.start()
+	// TODO (lruch): start go routine to monitor and update progress of running jobs
+	// TODO (lruch): add watchdogs to monitor defunct jobs
+}
+
+func (s *Scheduler) start() {
+	s.logger.Debug("starting scheduling loop")
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-s.baseContext.Done():
+			s.logger.Debug("scheduling loop shutting down")
+			return
+
+		case <-timer.C:
+			s.logger.Debug("waking up to run jobs")
+
+			repo, err := s.jobRepoFn()
+			if err != nil {
+				s.logger.Error("error creating job repo", "error", err)
+				break
+			}
+
+			runs, err := repo.RunJobs(s.baseContext, s.serverId, job.WithRunJobsLimit(s.runJobsLimit))
+			if err != nil {
+				s.logger.Error("error getting jobs to run from repo", "error", err)
+				break
+			}
+
+			for _, r := range runs {
+				err := s.runJob(r)
+				if err != nil {
+					s.logger.Error("error starting job", "error", err)
+					if _, inner := repo.FailRun(s.baseContext, r.PrivateId); inner != nil {
+						s.logger.Error("error updating failed job run", "error", inner)
+					}
+				}
+			}
+		}
+
+		s.logger.Debug("scheduling loop going back to sleep")
+		timer.Reset(s.runJobsInterval)
+	}
+}
+
+func (s *Scheduler) runJob(r *job.Run) error {
+	jobId := JobId(r.JobId)
+	j, ok := s.registeredJobs.Load(jobId)
+	job := j.(Job)
+	s.logger.Debug(fmt.Sprintf("starting job run %q for job (%v: %v)", r.PrivateId, jobId, job.Name()))
```

Comment:
Don't use printf and friends with the logger; use the logger's arguments. This would be something like:

`s.logger.Debug("starting job run", "id", r.PrivateId, "job_id", jobId, "name", job.Name())`

Also what's the difference between the private ID and the jobId here? Shouldn't they be the same?

Reply:
ty